### PR TITLE
Feature/rpc source fix

### PR DESCRIPTION
**Pull request by debruce**
some tracepoints to time the rpc code, and the initial rpc 'fix':

### CODE REVIEW

File: test/CMakeLists.txt
Comment (gregmedd) [inline test/CMakeLists.txt:42]:
```suggestion
find_package(Boost REQUIRED COMPONENTS interprocess)
```

File: lib/src/upZenohClient.cpp
Comment (gregmedd) [inline lib/src/upZenohClient.cpp:81]: Why is this even necessary? `w_handle` is a weak pointer - the intent is that it _automatically_ gets reset when all shared pointers to that instance are lost. Calling reset on this here _doesn't_ clean up any existing instances.

File: lib/src/upZenohClient.cpp
Comment (gregmedd) [inline lib/src/upZenohClient.cpp:55]: Instead of copy-pasting code, could this be streamlined with std::optional?

```suggestion
std::shared_ptr<UpZenohClient> UpZenohClient::instance() noexcept {
    return instance(std::nullopt_t);
}

std::shared_ptr<UpZenohClient> UpZenohClient::instance(std::optional<const uprotocol::v1::UAuthority&> src_authority) noexcept {
    static std::weak_ptr<UpZenohClient> w_handle;
```

File: lib/src/upZenohClient.cpp
Comment (gregmedd) [inline lib/src/upZenohClient.cpp:68]: If taking a std::optional for the source, then it would be checked here:
```suggestion
        if (!src_authority) {
            return nullptr;
        }
        auto rpc_handle = static_pointer_cast<ZenohRpcClient>(handle);
        rpc_handle->clientAuthority = *src_authority;
```

File: lib/src/zenohRpcClient.cpp
Comment (gregmedd) [inline lib/src/zenohRpcClient.cpp:165]: This looks like test code in the production code. What is this doing here?

File: lib/src/zenohUTransport.cpp
Comment (gregmedd) [inline lib/src/zenohUTransport.cpp:36]: Minor: There are lots of whitespace changes in this PR that aren't actually relevant to the change. Would it be very difficult to undo those to minimize the number of irrelevant changes?

File: lib/src/zenohUTransport.cpp
Comment (gregmedd) [inline lib/src/zenohUTransport.cpp:93]: Minor: Another example of a whitespace change that doesn't appear necessary

**Review by gregmedd — CHANGES_REQUESTED** [review]
I'd like to see the duplicated code for both flavors of `UpZenohClient::instance` removed and I would like to know more about why `UpZenohClient::purge()` was needed. Otherwise there are a few clean-up items needed.

Any comment starting with `Minor:` is not super important. I wouldn't hold up the PR over them, but those changes would be _nice to have_.

File: lib/src/zenohRpcClient.cpp
Comment (stevenhartley) [reply lib/src/zenohRpcClient.cpp:165]: handler is wrong. Response needs to be .setRpcResponse() which will populate the name, instance and ID correctly (if we are talking about the RPC response).

File: lib/src/upZenohClient.cpp
Comment (debruce) [reply lib/src/upZenohClient.cpp:68]: I intended to make the src_authority optional. What is is the point of returning nullptr if the optional is not provided? Seems better to just not make it optional. Or, am I misunderstanding your point?

File: lib/src/zenohRpcClient.cpp
Comment (debruce) [reply lib/src/zenohRpcClient.cpp:165]: I've changed to setRpcResponse(), but it doesn't compile.  When I look at the BuildResource.h, it doesn't seem to match expectations. Steven Hartley, can you clarify about this?

File: lib/src/zenohUTransport.cpp
Comment (debruce) [reply lib/src/zenohUTransport.cpp]: I'm trying to make changes to limit the whitespace differences, but it seems like a bit of a burden, given that I don't understand our whitespace policy, and we don't have a code formatter. Given that the code is fundamentally broken, and will need massive changes once we get past this, I would suggest that we put a priority on choosing a code formatter, having, and reformatting PR, and then insist that all PRs are code formatted with the agreed-upon formatter with every PR.

File: lib/src/upZenohClient.cpp
Comment (gregmedd) [reply lib/src/upZenohClient.cpp]: I think you might be missing the point. The source parameter is only required when constructing a new instance. When a valid instance already exists, it is optional (or really it's an error to provide a source when the instance is already initialized). Returning nullptr is the same error propagation strategy as when one of the lower layers fails to initialize (see a couple lines down in the file).

The goal here is to have one implementation of the instance function.

File: lib/src/zenohRpcClient.cpp
Comment (stevenhartley) [reply lib/src/zenohRpcClient.cpp]: > I've changed to setRpcResponse(), but it doesn't compile. When I look at the BuildResource.h, it doesn't seem to match expectations. Steven Hartley, can you clarify about this?

what do you mean? the API seems to be right....

File: lib/src/zenohUTransport.cpp
Comment (gregmedd) [reply lib/src/zenohUTransport.cpp]: The ask is simply not to change whitespace in sections of the code where you are not making actual code changes. No formatter should be needed to not change things. For example, the whitespace changes pointed out here are adding / removing blank lines between code lines.

Comment (gregmedd) [reply lib/src/zenohUTransport.cpp]: Looking at the latest files, I think we can call this resolved. Thanks for cleaning that up.

File: lib/include/up-client-zenoh-cpp/client/upZenohClient.h
Comment (gregmedd) [inline lib/include/up-client-zenoh-cpp/client/upZenohClient.h:51]: In this latest update, you've removed `static std::shared_ptr<UpZenohClient> instance () noexcept`. That will increase burden on users of this. The ask was simply to have that call this version with the optional parameters set to `std::nullopt_t`.

You can see the fallout of removing the interface in the scale of changes that were needed in all of the test files.

Additionally, the default parameters here are incorrect. _If_ using default parameters here, the should be `std::nullopt_t`.

Please apply one of these suggestions and roll back to the previous method for retrieving the instances in all of the tests:

```suggestion
            static std::shared_ptr<UpZenohClient> instance() noexcept {
                return instance(std::nullopt_t, std::nullopt_t);
            }

            static std::shared_ptr<UpZenohClient> instance(
                std::optional<uprotocol::v1::UAuthority> = uprotocol::v1::UAuthority(),
                std::optional<uprotocol::v1::UEntity> = uprotocol::v1::UEntity()) noexcept;
```

*OR*

```suggestion
            static std::shared_ptr<UpZenohClient> instance(
                std::optional<uprotocol::v1::UAuthority> = std::nullopt_t,
                std::optional<uprotocol::v1::UEntity> = std::nullopt_t) noexcept;
```

File: lib/src/upZenohClient.cpp
Comment (gregmedd) [inline lib/src/upZenohClient.cpp:56]: It's better to check for the existence of the source parameters before constructing an instance using make_shared. Additionally, I would recommend passing the source parameters through to the `ZenohRpcClient` as constructor parameters instead of relying on pointer casting and setting parameters inside of another class. Lastly, the check on line 51 will never fail as we do not enter this block when `w_handle.use_count() > 0`.

I recommend changing this code to this:

```suggestion
        // When constructing a new instance for the first time, these parameters must be provided.
        if (!src_authority || !src_entity) {
            return nullptr;
        }
        
        handle = std::make_shared<UpZenohClient>(ConstructToken(), *src_authority, *src_entity);
```

If you want to check that `src_authority` and `src_entity` are not set when an instance already exists, that check will need to be done right before line 39 and line 45.

File: lib/src/zenohUTransport.cpp
Comment (gregmedd) [reply lib/src/zenohUTransport.cpp]: Adding to this - the only substantive changes in this file are the comments that were added. Maybe this file should just be reverted to its state on main.

File: test/src/testPubSub.cpp
Comment (gregmedd) [inline test/src/testPubSub.cpp:279]: Revert this change. Set up and tear down tasks should be done in the `SetUp()`, `TearDown()`, `SetUpTestSuite()`, and `TearDownTestSuite()` methods in the `TestPubSub` class.

File: test/src/testPubSub.cpp
Comment (gregmedd) [inline test/src/testPubSub.cpp:64]: Is the intent to keep the same instance for every test case? I think it would be better to make an instance at the start of each test and clean it up at the end (using `SetUp()` and `TearDown()`).

File: test/src/testRpcClient.cpp
Comment (gregmedd) [inline test/src/testRpcClient.cpp:1]: Can we temporarily abandon this test and remove it from the CMakeLists.txt. It is in need of significant rewrite.

**Review by gregmedd — CHANGES_REQUESTED** [review]
A few more changes needed.

**Review by gregmedd — APPROVED** [review]
Looks good to me. Thanks for the hard work @debruce!

**Review by stevenhartley — COMMENTED** [review]
Merging to unblock build, code in general needs revisiting

**Pull request merged**
